spacy/cli/apply.py

```diff
@@ -0,0 +1,124 @@
+import tqdm
+import sys
+
+from ._util import app, Arg, Opt, setup_gpu, import_code
+from typing import Optional, Generator, Union
+from pathlib import Path
+from wasabi import msg
+
+from ..tokens import Doc, DocBin
+from ..vocab import Vocab
+from .. import util
+
+
+path_help = ("Location of the documents to predict on."
+             "Can be a single file in .spacy format or "
+             "a text file with one document per line."
+             "If a directory is provided each "
+             "text file in the directory will be treated "
+             "as a single document.")
+out_help = "Path where to save the result .spacy file"
+code_help = ("Path to Python file with additional "
+             "code (registered functions) to be imported")
+gold_help = "Use gold preprocessing provided in the .spacy files"
+
+
+def _stream_data(
+    data_path: Path,
+    vocab: Vocab,
+    suffix: Optional[str] = None
+) -> Generator[Union[str, Doc], None, None]:
+    """
+    Load data which is either in a single file
+    in .spacy or plain text format or multiple
+    text files in a directory. If a directory
+    is provided skip subdirectories and undecodeable
+    files.
+    """
```

Comment:
It should also be able to handle recursively searching a directory of `.spacy` files. Look at how `spacy convert` handles this?

Reply:
Ok I checked `walk_directory` thanks! I don't even accept multiple `.spacy` files at the moment, because I was wondering why would one have multiple of them. If we do the `walk_directory` then if the user provides a directory as input I will also require a `suffix`.

Reply:
Changed the behavior to either take a single file which can be plain text or `.spacy` file or a directory which then gets walked but then a `suffix` needs to be provided.

Reply:
I think that the existing `walk_directory` function may have too much convert-specific logic in it for use here. I'm not sure we've clearly defined what the accepted non-`.spacy` formats are? The PR description says one text per line but the code looks like it's treated each non-`.spacy` file as a single text? I also though there was some discussion about supporting `"text"` in JSONL dicts?

The pipeline can potentially take a mix of texts and `Doc`s, so it may not be necessary to provide a suffix.

Reply:
That's true that we could support a mix, but then we should say that okay, you have to have a directory where:

1. text files are called `.txt` and we treat each file as one document.
2. `jsonl` files are called `.jsonl` and we will read the "text" field.
3. you can have serialized `DocBin`s and they must be called `.spacy` also.

So I can change then the behavior by not having this option that one text file can have multiple documents, rather, we are always walking the directory that the user provided and we take the extensions at face value. How does this sound?

Reply:
To avoid too many restrictions in terms of filenames, I'd recommend:

1. `.spacy`
2. `.jsonl`
3. anything else is treated as text with one text per file